pkg/blockbuilder/blockbuilder.go

```diff
@@ -235,6 +240,61 @@ func (b *BlockBuilder) consumeJob(ctx context.Context, key schedulerpb.JobKey, s
 	return b.consumePartitionSection(ctx, logger, consumer, builder, spec.Partition, spec.StartOffset, spec.EndOffset)
 }
 
+type fetchPoller interface {
+	PollFetches(context.Context) kgo.Fetches
+}
+
+type fetchWrapper struct {
+	fetchers *ingest.ConcurrentFetchers
+}
+
+func (f *fetchWrapper) PollFetches(ctx context.Context) kgo.Fetches {
+	fetch, _ := f.fetchers.PollFetches(ctx)
+	return fetch
+}
+
+var _ fetchPoller = (*fetchWrapper)(nil)
+
+// newFetchers creates a new concurrent fetcher, retrying until it succeeds or the context is cancelled.
+// The returned error is the last error encountered.
+func (b *BlockBuilder) newFetchers(ctx context.Context, logger log.Logger, partition int32, startOffset int64) (*ingest.ConcurrentFetchers, error) {
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 5 * time.Second,
+		MaxRetries: 10,
+	})
+
+	var lastError error
+
+	for boff.Ongoing() {
+		f, ferr := ingest.NewConcurrentFetchers(
+			ctx,
+			b.kafkaClient,
+			logger,
+			b.cfg.Kafka.Topic,
+			partition,
+			startOffset,
+			b.cfg.Kafka.FetchConcurrencyMax,
+			100_000_000,
+			false,
+			5*time.Second,
+			nil, // Don't need a reader since we've provided the start offset.
+			ingest.NewGenericOffsetReader(func(context.Context) (int64, error) { return 0, nil }, 5*time.Second, logger), // Dummy.
```

Comment:
This is the thing that is still outstanding in this PR. If you read how it's used, it's used to figure out which offset to resume from _in the event_ that we encounter an `OffsetOutOfRange` error. We should not ever encounter that. If we did, we should probably abort the block-builder job. We do not want the reader to resume from "any old place" if there's an error. I'd rather influence the concurrentfetcher to return an error if this happened. I'll think about it, but open to discussion in the meantime.

Reply:
i'm not sure if starting at 0 also won't trigger OffsetOutOfRange if 0 has been truncated already and lead to some infinite error loop. I think that was the idea of the start offset reader.

Is there a point in trying to compact an offset that's been truncated? Technically that data no longer exists, so we might as well admit data loss and sound some alarms.